### PR TITLE
ISPN-5338 Extend JSR-107 tests

### PR DESCRIPTION
**Pull request by mcimbora**
* Added basic tests covering clustered scenarios. Test templates are defined in commons module, with 'embedded' and 'remote' modules providing context to run the tests. 
* For remote module, using jsr107 annotations has not been tested yet - waiting for @isavin 's feedback on whether it's supported - What interceptor should be specified in beans.xml when interceptor hierarchy resides in embedded module? Even doc is unclear about this, see http://infinispan.org/docs/7.2.x/user_guide/user_guide.html#_use_jcache_caching_annotations
* For now, 'testExpiration' test has been disabled in remote module as it fails constantly
* Migrated jcache module to JUnit
* Feel free to suggest more scenarios!

### CODE REVIEW

File: jcache/commons/src/test/java/org.infinispan.jcache/AbstractTwoCachesAnnotationsTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
+package org.infinispan.jcache;
```
Comment (galderz) [inline jcache/commons/src/test/java/org.infinispan.jcache/AbstractTwoCachesAnnotationsTest.java]: The folder where this file is located might be wrong? It shows as being `jcache/commons/src/test/java/org.infinispan.jcache` when it should be `jcache/commons/src/test/java/org/infinispan/jcache`

Comment (galderz) [inline jcache/commons/src/test/java/org.infinispan.jcache/AbstractTwoCachesAnnotationsTest.java]: Same thing happens with many of the other files. I've been bitten with this problem before and it causes problems with IDEs in particular.

**Comment by galderz** [comment]
What was the reason for moving the module over to `JUnit`? Although this is something we want to be doing down the line, we don't yet have a testing kit to support tests, so you've ended up avoiding all test cache manager creation utils and avoided extending any support classes that exist to help us test, e.g. check `core` tests to see what I mean. Also, the way the cache managers are constructed won't work fine if at some point the testsuite runs in paralell.

Unless there's a reason to not use `testng`, you should it and base your tests around the existing tools and support classes in order to avoid unnecessary code duplication and unwanted side effects with the rest of the testsuite. Our current `testng` setup has ways to avoid things leaking, and hence if not using it we have no safety nets.

**Comment by galderz** [comment]
Also, for remote tests, there's several base test classes in client/hotrod-client testsuite

**Comment by tristantarrant** [comment]
In fact, creating a JUnit base testkit is essential to the migration, and should be a separate project. Once that is ready it shouldn't be difficult to migrate all of the testsuite (possibly with some automated search/replace).

File: jcache/commons/src/test/java/org.infinispan.jcache/AbstractTwoCachesAnnotationsTest.java (hunk 1 above)
Comment (mcimbora) [inline jcache/commons/src/test/java/org.infinispan.jcache/AbstractTwoCachesAnnotationsTest.java]: Good catch!

**Comment by mcimbora** [comment]
@galderz I wanted to create hierarchy of tests, with AbstractXXXTests residing at the top to avoid code duplication. Remote tests were originally located in server/integration/testsuite (which uses JUnit), as I used full server to run them. With testng requiring you to extend org.jboss.arquillian.testng.Arquillian class when defining Arquillian tests, there was a problem making AbstractTwoCachesAnnotationsTest generic (without extending testng's Arquillian), as JCacheTwoCachesAnnotationsTest in embedded module would then need to extend 2 classes at the same time (Arquillian & AbstractTwoCachesAnnotationsTest). Being able to configure the tests easily with JUnit's annotations (RunWith), I switched to that, having in mind that testng -> junit migration is planned for future anyway. Once I migrated the tests to jcache.remote module I preferred to keep the behavior.

If we're not ready to use JUnit right away, I'll try to convert those tests back to testng. I'll post more informotation in this thread if I come across any limitations. Thanks!

**Comment by galderz** [comment]
@mcimbora Thanks for the explanation, that's a fair case for the approach you followed. Even if you use JUnit however, you should use the helpers available in TestCacheManagerFactory to create cache managers in embedded, and you should use TestingUtil for stopping/killing things in a safe way.

**Comment by galderz** [comment]
Also, instead of XML configurations, it's better to use programmatic configuration for tests since it keeps everything, both configuration and test code in the same place. XML configurations are preferred when either configuration is very complex or when testing XML parsing or similar.

**Comment by galderz** [comment]
Closing this, reopen when ready.

**Pull request closed without merging**